FEATURE: Neos 9.0 compatibility

mhsdesign opened the pull request:
# This change introduces support for Neos 9.

When starting the development it was eminent that not all features of NodeTemplates version 1.0 could be considered. To ease maintenance the [version 2 of NodeTemplates](https://github.com/Flowpack/Flowpack.NodeTemplates/releases/tag/2.0.0) was a complete rewrite for this package in anticipation of Neos 9. It removed incompatible features but also introduced new functionalities. The core functionality was rewritten to not depend on active records so it can be adjusted to issue a set of content repository commands.
That way it became possible to reuse the test-suite as well as parts of the core domain and be up-mergeable.
Multiple up-merges from version 2 have been successful and this strategy will allow us to fix bugs or ease introducing new minor features.

That being said all features of NodeTemplates version 2 are herby adjusted/refactored to work with Neos 9 while passing all tests:

## Core functionality, NodeTemplates via `options.template`

Full support of the feature set of version 2.
The deprecated EEL context variable `triggeringNode` was removed as [discussed](https://github.com/Flowpack/Flowpack.NodeTemplates/pull/70). Please use `site` or `parentNode` instead.

## `flow nodetemplate:validate`

Includes new `--site` option: the Neos site, which determines the content repository. Defaults to the first available one.

## `flow nodeTemplate:createFromNodeSubtree`

Includes new `--site` option: the Neos site, which determines the content repository. Defaults to the first available one.

## Additional notes

- introduce PHPStan Level 8
- run ci for PHP8.2 & PHP8.3
- align uri path segment generation to 9.0 `NodeCreationService::generateUriPathSegment` (see also [discussion to move it into the core](https://github.com/neos/neos-development-collection/pull/4324))
- The custom build declarative way to change the cr in version 2 via `Flowpack\NodeTemplates\Domain\NodeCreation\NodeMutator` was replaced by using native ESCR commands
- adjust fully to NodeType properties/references split, which is already partly part of the version 2 philosophy
- use public constructor promotion for `TransientNode` but not in other parts that should be up-mergeable (like `Template`)
- discuss how to calculate node aggregate ids of tethered nodes. We decided against making them deterministic [here](https://github.com/neos/neos-development-collection/pull/4313).
- Builds on the Neos.Ui node creation handler refactoring https://github.com/neos/neos-ui/pull/3519
- Requires this Neos 9 bugfix: https://github.com/neos/neos-development-collection/issues/4351
- Implements workaround for this Neos 9 "bug": https://github.com/neos/neos-development-collection/issues/5154 / see also https://github.com/Flowpack/Flowpack.NodeTemplates/issues/41

mhsdesign committed:
WIP: FEATURE: Neos 9.0 Prototype

mhsdesign committed:
FEATURE: First working draft. 1 functional Test passes ;)

this one ^^
--filter testNodeCreationMatchesSnapshot1

mhsdesign committed:
FEATURE: Add support for setting references

mhsdesign committed:
TASK: re-enable exception handling

mhsdesign committed:
TASK: ensure node has uri path segment

mhsdesign committed:
TASK: Speed up tests by running in synchronous mode, and use own testing `node_templates` cr

mhsdesign committed:
TASK: Re-enable other tests

mhsdesign committed:
TASK: Check for invalid node-constraints and update snapshots

mhsdesign committed:
TASK: Update snapshot to contain `null` property value

See https://github.com/neos/neos-development-collection/pull/4322 if this actually correct behavior or not

mhsdesign commented:
Some notes:

related: https://github.com/Flowpack/Flowpack.NodeTemplates/issues/50

related: https://github.com/Flowpack/Flowpack.NodeTemplates/issues/35#issuecomment-1550819560


references need special handling: https://github.com/neos/neos-ui/blob/4681fd003d9f9b6a14a47440d859265bad35044e/Classes/Domain/Model/Changes/Property.php#L152-L180



-------------


Question is, how to support usecases like #60 

```
            childNodesToCopy: "${q(node).find('#' + data.templateNodeIdentifier).children('main').children().get()}"
```

as flowquery will only work with a "real" node and we have one yet. Maybe we need a helper for this?

```
Neos.Node.findNodeById(id, subgraph)
```

edit: as discussed in https://github.com/Flowpack/Flowpack.NodeTemplates/pull/70#discussion_r1255520140 we will make the `parentNode` or `site` node available ;)

theilm commented:
> Question is, how to support usecases like #60

I think for this use case, the flowquery is not the problem: the nodes we try to get should be already materialized. The problem is: ATM we can copy the nodes when reacting to the signal; in the future, the target is not yet materialized and we have to modify the template based on the nodes we want to copy. Although this would be part of the package implementing the feature, we should think about how this could be done.

mhsdesign committed:
TASK: Adjust to NeosUI commands->first update and expose parentNode in context

mhsdesign committed:
Upmerge nodetemplates version 2 changes into version 3-dev

This upmerge includes changes until 2.0.1 aswell as the WIP features #58 and #67

mhsdesign committed:
TASK: Hackily fix test

mhsdesign committed:
Upmerge remote-tracking branch 'origin/feature/separateNodeCreationWithQueuedMutators' into neos9

mhsdesign commented on Classes/Domain/TemplateNodeCreationHandler.php:
```diff
@@ -42,32 +37,44 @@ class TemplateNodeCreationHandler implements NodeCreationHandlerInterface
         $evaluationContext = [
             'data' => $data,
-            'triggeringNode' => $node,
+            // todo evaluate which context variables
```
see https://github.com/Flowpack/Flowpack.NodeTemplates/pull/70#discussion_r1255520140

mhsdesign committed:
Merge remote-tracking branch 'origin/feature/separateNodeCreationWithQueuedMutators' into neos9

mhsdesign committed:
Merge remote-tracking branch 'origin/2.0' into neos9

mhsdesign committed:
TASK: Prepare tests for neos 9

mhsdesign committed:
TASK: Test patch neos stuff

mhsdesign committed:
TASK: Make nodeId deterministic for testing

mhsdesign commented on .github/workflows/tests.yml:
```diff
@@ -46,10 +58,30 @@ jobs:
+          composer require --no-update --no-interaction cweagans/composer-patches:^1.7.3
+          # the packages might have to be installed from source, so the patch works
+          patches='{
+            "neos/neos-ui": {
+              "improvedNodeCreationHandler": "https://github.com/neos/neos-ui/pull/3519.patch"
+            },
+            "neos/neos-development-collection": {
+              "tetheredNodeIdsDeterministic": "https://github.com/neos/neos-development-collection/pull/4313.patch"
+            }
+          }'
+          composer config extra.patches --json ${patches//[[:space:]]/}
```
Those are the patches you need ;)

mhsdesign committed:
TASK: Remove dependency to WIP branch and copy code

mhsdesign committed:
TASK: Adjust to changes in Neos

mhsdesign committed:
TASK: Fix generateUriPathSegment

mhsdesign committed:
TASK: Fix uriPathSegment if it depends on the title of the node creation data

fixes `/Features/Pages/2`

```
--- Expected
+++ Actual
@@ @@
     ],
     "properties": {
         "title": "Page1",
-        "uriPathSegment": "page1"
+        "uriPathSegment": "new-node"
     }
 }
 ```

mhsdesign committed:
Merge remote-tracking branch 'origin/2.0' into neos9

mhsdesign committed:
TASK: Adjust to changes in neos 9 dev (with phpstan)

mhsdesign committed:
TASK: Fix ci distribution installation

mhsdesign committed:
TASK: Use dev distribution in ci because sub splits are broken atm

mhsdesign committed:
TASK: Adjust unit and functional tests to changes in neos 9 dev

mhsdesign committed:
TASK: Use tetheredDescendantNodeAggregateIds instead of deterministic tethered node ids

The pr https://github.com/neos/neos-development-collection/pull/4313 was declined and tethered node ids cannot be calculated deterministically but must be generated and specified in the command beforehand.

mhsdesign committed:
Merge remote-tracking branch 'origin/2.0' into neos9

mhsdesign committed:
BUGFIX: Nested tethered nodes

mhsdesign committed:
TASK: Level up Phpstan to 6

mhsdesign committed:
Merge remote-tracking branch 'origin/2.0' into neos9

mhsdesign committed:
TASK: Level up ++ phpstan

mhsdesign committed:
TASK: Level up ++ phpstan to 8

mhsdesign committed:
TASK: Update to adjustments in Neos and Neos.Ui

- introduce TemplateNodeCreationHandlerFactory
- adjust to api changes in cr (NodeType::hasProperty)
- fix unit tests
- fix functional tests
 - add unstructured node and cr root

mhsdesign committed:
TASK: Simplify `ContentRepositoryTestTrait`

As the logic is not anymore in Neos, and we execute the tests synchronous

mhsdesign committed:
TASK: Adjust to changes in the NeosUi

mhsdesign committed:
Merge remote-tracking branch 'origin/2.0' into neos9

mhsdesign committed:
TASK: Adjust to changes in UI

mhsdesign committed:
FEATURE: Neos9-Beta10 compatibility

mhsdesign committed:
BUGFIX: Add workaround for unsetting properties with default values

See https://github.com/neos/neos-development-collection/issues/5154

mhsdesign committed:
TASK: Remove patches from CI after merge of 3519

mhsdesign commented:
The node creation handler overhaul was merged https://github.com/neos/neos-ui/pull/3519 and the node was made workspace aware https://github.com/neos/neos-development-collection/pull/5042, the biggest changes in Neos 9 that would affect this package are done and the main functionality of this package is now fully compatible (without patches) with the latest beta Neos9-Beta10.
That means we can merge this into a target branch wed like to eventually tag a first V3.0 some time.
~We can merge into `master` (though its behind 2.0) or create a new `main` branch, or create just a 3.0 branch. Your choice :D~ Edit lets do 3.0

mhsdesign committed:
Merge remote-tracking branch 'origin/2.0' into neos9

mhsdesign committed:
TASK: Introduce `site` variable in Neos 9

Adjustments for https://github.com/Flowpack/Flowpack.NodeTemplates/pull/70

mhsdesign committed:
FEATURE: Neos9 support for `flow nodeTemplate:validate`

Introduced via https://github.com/Flowpack/Flowpack.NodeTemplates/pull/58

mhsdesign committed:
TASK: Unskip test after merge of core bugfix

see https://github.com/neos/neos-development-collection/issues/4351

mhsdesign committed:
FEATURE: Migrate `flow nodeTemplate:createFromNodeSubtree` to Neos9

introduced with https://github.com/Flowpack/Flowpack.NodeTemplates/pull/40

mhsdesign committed:
TASK: Use FlowEntitiesTrait::truncateAndSetupFlowEntities to reset persistence

Previously the doctrine entities would still exist of the previous test.

mhsdesign committed:
TASK: Run tests also on PHP8.3

mhsdesign committed:
TASK: Hotfix for transaction errors in CI

Neos.ContentRepository.Core/Classes/EventStore/EventPersister.php:47

RuntimeException: A transaction is active already, can't commit events!

-----------

and succeeding errors:

Neos.ContentRepository.Core/Classes/Infrastructure/DbalCheckpointStorage.php:88

RuntimeException: Failed to acquire checkpoint lock for subscriber "Neos\ContentRepository\Core\Projection\ContentStream\ContentStreamProjection" because a transaction is active already

mhsdesign committed:
TASK: Improve documentation of `nodeTemplate` CLI commands

mhsdesign committed:
TASK: Small cosmetic adjustment

mhsdesign committed:
Merge remote-tracking branch 'origin/2.0' into neos9

mhsdesign committed:
TASK: Improve DI for `TemplateNodeCreationHandler` and early return if `$$nodeType` does not exist like in v2

Commit:
Adjust to Neos9 Beta15

Commit:
Adjust tests to beta15

Commit:
Adjust to beta 16

mhsdesign committed:
TASK: Adjust testing setup to boot after content repository changes

mhsdesign committed:
TASK: Avoid issuing `SetNodeProperties` if its a noop

> The command "SetNodeProperties" for node 059f8bc9-e262-4f5d-a1e5-50ccf0f74d87 must contain property values

mhsdesign committed:
TASK: Update unit tests to adjusted node type manager constructor

mhsdesign committed:
TASK: Leverage new `Commands` class instead of passing `NodeCreationCommands` recursively

Requires https://github.com/neos/neos-ui/pull/3925

mhsdesign committed:
TASK: Use new UI warning

https://github.com/neos/neos-ui/pull/3933

mhsdesign committed:
TASK: Update `actions/cache@v3`

mhsdesign committed:
TASK: Fix CI `cr:setup` does not work because nodetypes are not initialised

nezaniel reviewed (approved):
lgtm and PHPStan level 8 4tw
